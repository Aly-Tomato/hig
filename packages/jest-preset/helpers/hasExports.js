module.exports = function hasExports(index, cases) {
  cases.forEach(({ name, value }) => {
    it(`exports ${name}`, () => {
      expect(index).toHaveProperty(name, value);
    });
  });
};
