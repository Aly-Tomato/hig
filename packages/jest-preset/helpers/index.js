/* eslint-disable global-require */
module.exports = {
  hasExports: require("./hasExports"),
  lastCallOfMock: require("./lastCallOfMock"),
  takeSnapshotsOf: require("./takeSnapshotsOf")
};
