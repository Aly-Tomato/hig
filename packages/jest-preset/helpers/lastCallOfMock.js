module.exports = function lastCallOfMock(mockFn) {
  return mockFn.mock.calls[mockFn.mock.calls.length - 1];
};
