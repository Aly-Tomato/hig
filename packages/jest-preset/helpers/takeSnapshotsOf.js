const React = require("react");
const renderer = require("react-test-renderer");

module.exports = function takeSnapshotsOf(Component, cases) {
  cases.forEach(({ desc, props }) => {
    it(desc, () => {
      const presenter = React.createElement(Component, props);
      const tree = renderer.create(presenter).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });
};
