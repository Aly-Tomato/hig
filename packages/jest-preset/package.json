{
  "private": true,
  "name": "@hig/jest-preset",
  "version": "0.1.0",
  "dependencies": {
    "babel-core": "^6.26.3",
    "babel-jest": "^23.4.2",
    "enzyme": "^2.9.1",
    "jest-enzyme": "^3.8.3",
    "jest-extended": "^0.6.1",
    "raf": "^3.4.0",
    "react-test-renderer": "^15.6.2"
  },
  "eslintConfig": {
    "extends": "@hig"
  },
  "jest": {
    "preset": "@hig/jest-preset"
  }
}
