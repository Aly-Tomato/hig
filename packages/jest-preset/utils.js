require("@hig/utils").generateId = jest.fn(key => `${key}-1`);
