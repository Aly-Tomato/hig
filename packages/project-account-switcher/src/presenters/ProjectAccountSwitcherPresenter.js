import React, { Component } from "react";
import PropTypes from "prop-types";
import Flyout from "@hig/flyout";
import "@hig/flyout/build/index.css";
import icons from "../../../icons/build";
import { anchorPoints } from "../../../flyout/src/anchorPoints";
import "./ProjectAccountSwitcherPresenter.scss";

/* eslint-disable jsx-a11y/no-noninteractive-element-interactions */
/* eslint-disable jsx-a11y/no-static-element-interactions */
/* eslint-disable react/no-danger */
/* eslint-disable object-shorthand */
/* eslint-disable no-unused-vars */

export default class ProjectAccountSwitcherPresenter extends Component {
  static constructPlaceholder(label) {
    return label.match(/\b(\w)/g).join("");
  }

  static propTypes = {
    /** Heading title for the list of Accounts */
    accountTitle: PropTypes.string,
    /** List of Accounts */
    accounts: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.string,
        image: PropTypes.string,
        label: PropTypes.string
      })
    ),
    /** Currently selected Account */
    activeAccount: PropTypes.shape({
      id: PropTypes.string,
      image: PropTypes.string,
      label: PropTypes.string
    }),
    /** Label for selected Accounts and Projects */
    activeLabel: PropTypes.string,
    /** Currently selected Project */
    activeProject: PropTypes.shape({
      id: PropTypes.string,
      image: PropTypes.string,
      label: PropTypes.string
    }),
    /** Handles Account selection, passed from Behavior */
    onAccountClick: PropTypes.func,
    /** Handles Project selection, passed from Behavior */
    onProjectClick: PropTypes.func,
    /** Handles Flyout display, passed from Behavior */
    onTargetClick: PropTypes.func,
    /** Menu heading title for the list of Projects */
    open: PropTypes.bool,
    /** Heading title for the list of Projects */
    projectTitle: PropTypes.string,
    /** List of Projects */
    projects: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.string,
        image: PropTypes.string,
        label: PropTypes.string
      })
    )
  };

  static defaultProps = {
    accountTitle: "Accounts",
    open: false,
    projectTitle: "Projects"
  };

  constructLabel() {
    const labels = [];
    if (this.props.activeAccount) {
      labels.push(this.props.activeAccount.label);
    }
    if (this.props.activeProject) {
      labels.push(this.props.activeProject.label);
    }
    return labels.join(" / ");
  }

  // this method constructs the image placeholder for the combination of
  // the current activeAccount and/or activeProject,
  // in lieu of an image for either
  constructLabelPlaceholder() {
    const placeholders = [];
    if (this.props.activeAccount) {
      placeholders.push(
        ProjectAccountSwitcherPresenter.constructPlaceholder(
          this.props.activeAccount.label
        )
      );
    }
    if (this.props.activeProject) {
      placeholders.push(
        ProjectAccountSwitcherPresenter.constructPlaceholder(
          this.props.activeProject.label
        )
      );
    }
    return placeholders.join("/");
  }

  accountsList() {
    return this.props.accounts.map(account => (
      /* eslint-list-disable-next-line jsx-a11y/no-noninteractive-element-interactions */
      <li
        className="project-account-switcher__item
                      project-account-switcher__item--account
                      project-account-switcher__item--active"
        data-account-id={account.id}
        onClick={this.props.onAccountClick}
      >
        <span className="project-account-switcher__item__image-wrapper">
          <span
            className="project-account-switcher__item__image-placeholder"
            data-account-id={account.id}
          >
            {ProjectAccountSwitcherPresenter.constructPlaceholder(
              account.label
            )}
          </span>
        </span>
        <span
          className="project-account-switcher__item__label"
          data-account-id={account.id}
        >
          {account.label}
        </span>
      </li>
    ));
  }

  projectsList() {
    return this.props.projects.map(project => (
      /* eslint-list-disable-next-line jsx-a11y/no-noninteractive-element-interactions */
      <li
        className="project-account-switcher__item
                      project-account-switcher__item--project
                      project-account-switcher__item--active"
        data-project-id={project.id}
        onClick={this.props.onProjectClick}
      >
        <span className="project-account-switcher__item__image-wrapper">
          <span
            className="project-account-switcher__item__image-placeholder"
            data-project-id={project.id}
          >
            {ProjectAccountSwitcherPresenter.constructPlaceholder(
              project.label
            )}
          </span>
        </span>
        <span
          className="project-account-switcher__item__label"
          data-project-id={project.id}
        >
          {project.label}
        </span>
      </li>
    ));
  }

  flyoutContent() {
    return (
      <div className="project-account-switcher__lists">
        {this.props.accounts.length > 0 && (
          <ul className="project-account-switcher__list">
            <span className="project-account-switcher__list__title">
              {this.props.accountTitle}
            </span>
            {this.accountsList()}
          </ul>
        )}
        {this.props.projects.length > 0 && (
          <ul className="project-account-switcher__list">
            <span className="project-account-switcher__list__title">
              {this.props.projectTitle}
            </span>
            {this.projectsList()}
          </ul>
        )}
      </div>
    );
  }

  componentHasNoLists() {
    const { accounts = [], projects = [] } = this.props;

    return !accounts.length && !projects.length;
  }

  render() {
    if (this.componentHasNoLists()) {
      console.warn("You must provide a list of accounts or projects");
      return null;
    }

    const activeLabel = this.props.activeLabel
      ? this.props.activeLabel
      : this.constructLabel();
    return (
      /* eslint-list-disable-next-line react/no-danger */
      <div
        className="project-account-switcher"
        onClick={this.props.onTargetClick}
      >
        <Flyout
          anchorPoint={anchorPoints.TOP_RIGHT}
          open={this.props.open}
          panel={({ innerRef }) => (
            <Flyout.Panel innerRef={innerRef}>
              <div className="project-account-switcher--custom-width-and-padding">
                {this.flyoutContent()}
              </div>
            </Flyout.Panel>
          )}
        >
          <div className="project-account-switcher__target">
            <div className="project-account-switcher__item">
              <span className="project-account-switcher__item__image-wrapper">
                <img
                  className="project-account-switcher__item__image"
                  alt={activeLabel}
                  src={""}
                />
                <span className="project-account-switcher__item__image-placeholder">
                  {this.constructLabelPlaceholder()}
                </span>
              </span>
              <span className="project-account-switcher__item__label">
                {activeLabel}
              </span>
            </div>
            <span className="project-account-switcher__target__caret">
              <div
                className="hig__icon hig__icon--24-size"
                dangerouslySetInnerHTML={{
                  __html: icons["caret-24"]
                }}
              />
            </span>
          </div>
        </Flyout>
      </div>
    );
  }
}
