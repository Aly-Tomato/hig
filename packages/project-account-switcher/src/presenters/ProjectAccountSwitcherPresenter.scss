.hig__project-account-switcher__target {
  display: flex;
  align-items: center;
  cursor: pointer;
}

.hig__project-account-switcher__target__item {
  padding-left: 10px;
}

.hig__project-account-switcher__target__caret {
  margin: 0 2px;
}
