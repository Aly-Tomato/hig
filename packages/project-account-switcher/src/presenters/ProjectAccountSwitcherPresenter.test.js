import { takeSnapshotsOf } from "@hig/jest-preset/helpers";
import ProjectAccountSwitcherPresenter from "./ProjectAccountSwitcherPresenter";

describe("project-account-switcher/presenters/ProjectAccountSwitcherPresenter", () => {
  beforeAll(() => {
    console.warn = jest.fn();
  });

  takeSnapshotsOf(ProjectAccountSwitcherPresenter, [
    {
      description: "renders without props",
      props: {}
    },
    {
      description: "renders with all props",
      props: {
        accountTitle: "Accounts",
        accounts: [
          { id: "1", label: "Account 1" },
          { id: "2", label: "Account 2" }
        ],
        activeAccount: { id: "1", label: "Account 1" },
        activeProject: { id: "2", label: "Project 2" },
        onAccountClick: function onAccountClick() {},
        onProjectClick: function onProjectClick() {},
        projectTitle: "Projects",
        projects: [
          { id: "1", label: "Project 1" },
          { id: "2", label: "Project 2" }
        ]
      }
    }
  ]);
});
