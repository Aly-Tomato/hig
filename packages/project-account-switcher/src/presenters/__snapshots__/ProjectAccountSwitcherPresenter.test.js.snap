// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`project-account-switcher/presenters/ProjectAccountSwitcherPresenter  1`] = `null`;

exports[`project-account-switcher/presenters/ProjectAccountSwitcherPresenter  2`] = `
<div
  className="hig__project-account-switcher__target"
  onClick={[Function]}
  onKeyDown={[Function]}
  role="button"
  tabIndex="0"
>
  <div
    className="hig__project-account-switcher__target__item hig__project-account-switcher__item"
  >
    <span
      className="hig__project-account-switcher__item__image-wrapper"
    >
      <img
        alt="Foo"
        className="hig__project-account-switcher__item__image"
        src=""
      />
      <span
        className="hig__project-account-switcher__item__image-placeholder"
      >
        A1/P2
      </span>
    </span>
    <span
      className="hig__project-account-switcher__item__label"
    >
      Foo
    </span>
  </div>
  <div
    className="hig__project-account-switcher__target__caret"
  >
    <div
      className="hig__icon hig__icon--24-size"
      dangerouslySetInnerHTML={
        Object {
          "__html": "<svg width=\\"10\\" height=\\"6\\" viewBox=\\"0 0 10 6\\" xmlns=\\"http://www.w3.org/2000/svg\\" role=\\"img\\" title=\\"caret-24\\"><polyline transform=\\"rotate(45 9.828 3)\\" points=\\"8 2 8 8 2 8\\" stroke=\\"#2E4258\\" stroke-width=\\"1.15\\" fill=\\"none\\" fill-rule=\\"evenodd\\" stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\"/></svg>",
        }
      }
    />
  </div>
</div>
`;
